src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseCodeReviewFeedbackService.java

```diff
@@ -0,0 +1,239 @@
+     * @param programmingExercise the associated programming exercise.
+     */
+    private ProgrammingExerciseStudentParticipation setIndividualDueDateAndLockRepository(ProgrammingExerciseStudentParticipation participation,
+            ProgrammingExercise programmingExercise, boolean invalidatePreviousResults) {
+        // The participations due date is a flag showing that a feedback request is sent
+        participation.setIndividualDueDate(now());
+
+        participation = programmingExerciseStudentParticipationRepository.save(participation);
+        // Circumvent lazy loading after save
+        participation.setParticipant(participation.getParticipant());
+        programmingExerciseParticipationService.lockStudentRepositoryAndParticipation(programmingExercise, participation);
+
+        if (invalidatePreviousResults) {
+            var participationResults = participation.getResults();
+            participationResults.forEach(participationResult -> participationResult.setRated(false));
+            this.resultRepository.saveAll(participationResults);
+        }
+
+        return participation;
+    }
+
+    /**
+     * Removes the individual due date for a participation. If the due date for an exercise is empty or is in the future, unlocks the repository,
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private void unlockRepository(ProgrammingExerciseStudentParticipation participation, ProgrammingExercise programmingExercise) {
+        if (programmingExercise.getDueDate() == null || now().isBefore(programmingExercise.getDueDate())) {
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation(participation);
+            participation.setIndividualDueDate(null);
+            this.programmingExerciseStudentParticipationRepository.save(participation);
+        }
+    }
+
+    private void checkRateLimitOrThrow(ProgrammingExerciseStudentParticipation participation) {
+
+        List<Result> athenaResults = participation.getResults().stream().filter(result -> result.getAssessmentType() == AssessmentType.AUTOMATIC_ATHENA).toList();
+
+        long countOfAthenaResultsInProcessOrSuccessful = athenaResults.stream().filter(result -> result.isSuccessful() == null || result.isSuccessful() == Boolean.TRUE).count();
+
+        long countOfSuccessfulRequests = athenaResults.stream().filter(result -> result.isSuccessful() == Boolean.TRUE).count();
+
+        if (countOfAthenaResultsInProcessOrSuccessful >= 3) {
+            throw new BadRequestAlertException("Cannot send additional AI feedback requests now. Try again later!", "participation", "preconditions not met");
+        }
+        if (countOfSuccessfulRequests >= 3) {
+            throw new BadRequestAlertException("Maximum number of AI feedback requests reached.", "participation", "preconditions not met");
+        }
+    }
```

Comment:
- It seems to me like the differentiation between the 2 if statements is very minor. I would expect that the first one checks if there is any in progress request and cancel if that is the case, not if the number of successful + in progress requests is >= 3
- As it stands now you allow the student to have 2 requests to Athena running at the same time. Is this intended?

Reply:
Two concurrent Athena requests cannot happen, as the individual due date is set.